frontend/app/src/components/home/LinkPreviews.svelte

```diff
@@ -0,0 +1,197 @@
+    export let me: boolean;
+
+    let list: HTMLElement | null | undefined = undefined;
+    let previousLinks: string[] = [];
+    let previews: Preview[] = [];
+
+    $: {
+        if (!arraysAreEqual(previousLinks, links)) {
+            previews = links.map(buildPreview);
+            previousLinks = links;
+        }
+    }
+
+    function arraysAreEqual(a: string[], b: string[]) {
+        if (a.length !== b.length) {
+            return false;
+        }
+
+        for (let i = 0; i < a.length; i++) {
+            if (a[i] !== b[i]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    function buildPreview(url: string): Preview {
+        let regexMatch = url.match(client.youtubeRegex());
+        if (regexMatch) {
+            return {
+                kind: "youtube",
+                url,
+                regexMatch,
+            };
+        }
+
+        regexMatch = url.match(client.twitterLinkRegex());
+        if (regexMatch) {
+            return {
+                kind: "twitter",
+                url,
+                tweetId: regexMatch[3],
+            };
+        }
+
+        return {
+            kind: "generic",
+            url,
+        };
+    }
+
+    function closestAncestor(
+        el: HTMLElement | null | undefined,
+        selector: string,
+    ): HTMLElement | null | undefined {
+        while (el) {
+            if (el.matches(selector)) {
+                return el;
+            }
+            el = el.parentElement;
+        }
+        return null;
+    }
+
+    function renderPreview(url: string): void {
+        for (const preview of previews) {
+            if (preview.url === url && preview.container) {
+                preview.container.style.setProperty("display", "flex");
+                break;
+            }
+        }
+    }
+
+    function adjustScroll(wrapper: HTMLElement) {
+        // if we are using reverse scroll rendering there is no need to adjust the scroll top when rendering previews
+        if (reverseScroll || $lowBandwidth) return;
+
+        list = list || closestAncestor(wrapper, ".scrollable-list");
+        if (list) {
+            list.scrollTop = list.scrollTop + wrapper.offsetHeight;
+        }
+    }
+
+    function removePreview(preview: Preview | undefined) {
+        if (preview) {
+            dispatch("remove", preview.url);
+        }
+    }
+</script>
+
+{#each previews as preview (preview.url)}
+    <div
+        class="preview"
+        bind:this={preview.container}
+        class:visible={preview.kind !== "generic"}
+        class:me>
+        {#if me}
+            <div class="remove" class:rtl={$rtlStore} on:click={() => removePreview(preview)}>
+                <CloseIcon viewBox="0 0 24 24" size={$iconSize} color={"var(--button-txt)"} />
+            </div>
+        {/if}
+        <div class="inner">
+            {#if preview.kind === "twitter"}
+                <Tweet tweetId={preview.tweetId} {intersecting} />
```

Comment:
So we should still have `on:rendered` against the `Tweet` component now shouldn't we?

Reply:
Oops - put it on YouTubePreview by mistake!